extension/inet/inet_escape_functions.cpp

```diff
@@ -0,0 +1,373 @@
+}
+
+enum class MatchType {
+	Unknown, // 0
+	CharRef, // 1 &<charref>
+	Decimal, // 2 &#<dec_val>
+	Hex      // 3 &#x<hex_val>
+};
+
+struct MatchLengthCounter {
+	using RESULT_TYPE = idx_t;
+
+	static void Operation(const char *input_data, idx_t start, idx_t count, RESULT_TYPE &result) {
+		result += count;
+	}
+	static void Operation(const char *input_data, RESULT_TYPE &result) {
+		Operation(input_data, 0, strlen(input_data), result);
+	}
+};
+
+struct MatchWriter {
+	using RESULT_TYPE = char *;
+
+	static void Operation(const char *input_data, idx_t start, idx_t count, RESULT_TYPE &result) {
+		memcpy(result, input_data + start, count);
+		result += count;
+	}
+	static void Operation(const char *input_data, RESULT_TYPE &result) {
+		Operation(input_data, 0, strlen(input_data), result);
+	}
+};
+
+struct Matcher {
+	template <class OP>
+	static void Decode(string_t &input, UnescapeBindData &info, typename OP::RESULT_TYPE &result) {
+		auto input_data = input.GetData();
+		auto input_size = input.GetSize();
+		uint32_t num = 0; // the interpretation of the input string to the actual number
+		idx_t start = 0;  // match's starts
+		MatchType type = MatchType::Unknown;
+		for (idx_t i = 0; i < input_size; ++i) {
+			if (type != MatchType::Unknown) {
+				if (input_data[i] == '&') { // the beginning of a new match
+					DecodeMatch<OP>(type, input_data, start, i - 1, num, result, info);
+					num = 0;
+					type = GetMatchType(input, i, start);
+				} else if (input_data[i] == ';') { // match ends with semicolon
+					DecodeMatch<OP>(type, input_data, start, i, num, result, info);
+					type = MatchType::Unknown;
+				} else { // check if the digit is valid for decoding
+					bool is_valid = true;
+					if (type == MatchType::Decimal) {
+						is_valid = CheckDecDigit(input_data, i, num);
+					} else if (type == MatchType::Hex) {
+						is_valid = CheckHexDigit(input_data, i, num);
+					}
+					if (!is_valid) {
+						if (start == i) {
+							DecodeMatch<OP>(MatchType::Unknown, input_data, start - idx_t(type), i, 0, result, info);
+						} else {
+							DecodeMatch<OP>(type, input_data, start, i - 1, num, result, info);
+							OP::Operation(input_data, i, 1, result);
+						}
+						type = MatchType::Unknown;
+					}
+				}
+			} else {
+				// declare a match
+				num = 0;
+				type = GetMatchType(input, i, start);
+				if (type == MatchType::Unknown) { // append the character in the result
+					OP::Operation(input_data, i, 1, result);
+				}
+			}
+		}
+		if (type != MatchType::Unknown) {
+			DecodeMatch<OP>(type, input_data, start, input_size - 1, num, result, info);
+		}
+	}
+
+	static MatchType GetMatchType(string_t input, idx_t &i, idx_t &start) {
+		MatchType type;
+		auto input_data = input.GetData();
+		auto input_size = input.GetSize();
+		auto num_of_digits = input_size - i;             // number of digits after i index
+		if (num_of_digits > 2 && input_data[i] == '&') { // the beggining of a match
```

Comment:
Nit: if we invert this we can do an early out, dealing with the inverse of this condition first, returning MatchType::Unknown directly